### PR TITLE
[github-action] build jobs failure due to deprecated macOS-12 environment

### PR DESCRIPTION
**Pull request by ZhangLe2016**
The macOS-12 environment is deprecated, consider switching to macOS-13, macOS-14 (macos-latest) or macOS-15.

For more details, see https://github.com/actions/runner-images/issues/10721

In addition, removed the -s option because the -s option is not supported by the realpath command in macOS 13 (Ventura) and later versions, and change the `swig@4` to `swig` due to no available formula with the name "swig@4" in macOS13.

### CODE REVIEW

File: android/build-commissioner-libs.sh [hunk 1]
```diff
@@ -27,7 +27,7 @@
 #  POSSIBILITY OF SUCH DAMAGE.
 #
 
-readonly CUR_DIR="$(dirname "$(realpath -s "$0")")"
+readonly CUR_DIR="$(dirname "$(realpath "$0")")"
```
[the remark this answers is not in the document]
Comment (ZhangLe2016) [reply android/build-commissioner-libs.sh:30]: yes, the job will failed by

"realpath: illegal option -- s"

File: tests/integration/common.sh [hunk 2]
```diff
@@ -29,7 +29,7 @@
 
 ## This file defines constants and common functions for test cases.
 
-readonly CUR_DIR=$(dirname "$(realpath -s $0)")
+readonly CUR_DIR=$(dirname "$(realpath $0)")
```
Comment (wgtdkp) [inline tests/integration/common.sh:32]: Is this change necessary?

File: script/bootstrap.sh [hunk 3]
```diff
@@ -138,7 +138,7 @@ else
     exit 1
 fi
 
-readonly CUR_DIR="$(dirname "$(realpath -s "$0")")"
+readonly CUR_DIR="$(dirname "$(realpath "$0")")"
```
Comment (wgtdkp) [inline script/bootstrap.sh:141]: Is this change necessary?

Comment (ZhangLe2016) [reply script/bootstrap.sh]: yes, for solving the following build problem
https://github.com/openthread/ot-commissioner/actions/runs/12156344538/job/33899828836

File: android/build-commissioner-libs.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (ZhangLe2016) [reply android/build-commissioner-libs.sh]: yes, it is for fixing the following build error
https://github.com/openthread/ot-commissioner/actions/runs/12156872228/job/33901451549

File: tests/integration/common.sh (hunk 2 above)
Comment (ZhangLe2016) [reply tests/integration/common.sh]: verified, no need to change it.

File: android/build-commissioner-libs.sh [hunk 4]
```diff
@@ -27,7 +27,13 @@
 #  POSSIBILITY OF SUCH DAMAGE.
 #
 
-readonly CUR_DIR="$(dirname "$(realpath -s "$0")")"
+if [ "$(uname)" = "Linux" ]; then
+    echo "OS is Linux"
+    readonly CUR_DIR="$(dirname "$(realpath -s "$0")")"
```
Comment (wgtdkp) [inline android/build-commissioner-libs.sh:32]: Looking at the "-s" option of the realpath command, I think it's okay to remove "-s" on both Linux and Macos.

```
➜  ~ realpath --help
Usage: realpath [OPTION]... FILE...
Print the resolved absolute file name;
all but the last component must exist

  -e, --canonicalize-existing  all components of the path must exist
  -m, --canonicalize-missing   no path components need exist or be a directory
  -L, --logical                resolve '..' components before symlinks
  -P, --physical               resolve symlinks as encountered (default)
  -q, --quiet                  suppress most error messages
      --relative-to=DIR        print the resolved path relative to DIR
      --relative-base=DIR      print absolute paths unless paths below DIR
  -s, --strip, --no-symlinks   don't expand symlinks
  -z, --zero                   end each output line with NUL, not newline
      --help        display this help and exit
      --version     output version information and exit
```

Comment (ZhangLe2016) [reply android/build-commissioner-libs.sh]: done

File: script/bootstrap.sh [hunk 5]
```diff
@@ -130,15 +129,12 @@ elif [ "$(uname)" = "Darwin" ]; then
         brew unlink cmake
         brew install cmake --HEAD
     }
-
-    ## Install coreutils for realpath
-    brew install coreutils
```
Comment (wgtdkp) [inline script/bootstrap.sh:135]: Is this intentional? Does "realpath" come with macos 13?

**Review by wgtdkp — APPROVED** [review]

File: script/bootstrap.sh (hunk 5 above)
Comment (ZhangLe2016) [reply script/bootstrap.sh:135]: yes, it is an intended behavior.  The `coreutils` has been installed in following `brew install xxx` command in line113. we no need to reinstall it again after changing `swig@4` to `swig`.

```
 brew install coreutils \
                 readline \
                 cmake \
                 ninja \
                 swig  \
                 lcov && true
```

File: .github/workflows/android-app-build.yml [hunk 6]
```diff
@@ -37,7 +37,7 @@ jobs:
       matrix:
         android-api: [26, 34]
         android-abi: [x86_64]
-        os: [macos-12, ubuntu-22.04]
+        os: [macos-13, ubuntu-22.04]
```
Comment (jwhui) [inline .github/workflows/android-app-build.yml:40]: Did you try `macos-14`?

Comment (ZhangLe2016) [reply .github/workflows/android-app-build.yml:40]: macOS-14 may introduce a new issue: 'Timeout waiting for emulator to boot.' Currently, there is no known solution to address this issue.

https://github.com/openthread/ot-commissioner/actions/runs/12174524147?pr=308

Comment (jwhui) [reply .github/workflows/android-app-build.yml:40]: Thanks for checking. I'm fine with deferring change to `macos-14` to later.
